cpp/pipelines/add_authority_external_ref.cc

```diff
@@ -18,120 +19,201 @@
+    auto duration = std::chrono::duration_cast<std::chrono::seconds>(load_file_end - load_file_start);
 
+    std::cout << std::endl
+              << "Total GND-ID: " << total_numbers_of_gnd_id_generated << std::endl
+              << "Total GND with Wikidata: " << total_number_of_wikidata << std::endl
+              << "Total GND with Wikipedia: " << total_number_of_wikipedia << std::endl;
+    std::cout << "Total time of computation: " << duration.count() << " second(s)" << std::endl;
 
+    input_file.close();
+    return true;
+}
 void ParseGndWikidataMappingFile(std::string filename,
-                                 std::unordered_map<std::string, std::vector<std::string>>* const gnd_to_wikidataid_and_wikipedia_link) {
-    std::ifstream file(filename);
-    if (file.is_open()) {
-        std::string line;
-        std::string act_gnd;
-        std::string act_wikidata;
-        std::string act_wikipedia;
-        while (std::getline(file, line)) {
-            const std::string NAME = "Name:";
-            const std::string GND = "GND:";
-            const std::string WIKIDATA = "Wikidata:";
-            const std::string WIKIPEDIA = "Wikipedia:";
-            if (StringUtil::StartsWith(line, NAME) and StringUtil::Contains(line, GND) and StringUtil::Contains(line, WIKIDATA)
-                and StringUtil::Contains(line, WIKIPEDIA))
-            {
-                act_gnd = line.substr(line.find(GND) + GND.length());
-                act_gnd = act_gnd.substr(0, act_gnd.find(WIKIDATA));
-                act_wikidata = line.substr(line.find(WIKIDATA) + WIKIDATA.length());
-                act_wikidata = act_wikidata.substr(0, act_wikidata.find(WIKIPEDIA));
-                act_wikipedia = line.substr(line.find(WIKIPEDIA) + WIKIPEDIA.length());
-                std::vector<std::string> wiki_elements = { StringUtil::TrimWhite(act_wikidata), StringUtil::TrimWhite(act_wikipedia) };
-                gnd_to_wikidataid_and_wikipedia_link->emplace(StringUtil::TrimWhite(act_gnd), wiki_elements);
-            }
+                                 std::unordered_map<std::string, std::vector<std::string>> * const gnd_to_wikidataid_and_wikipedia_link) {
+    try {
+        std::vector<std::vector<std::string>> lines;
+        std::vector<std::string> wiki_elements;
+        TextUtil::ParseCSVFileOrDie(filename, &lines, ';', '"');
+        for (const auto &line : lines) {
+            // gnd = line[0], wikidata = line[1], wikipedia = line[2]
+            wiki_elements = { StringUtil::TrimWhite(line.at(1)), StringUtil::TrimWhite(line.at(2)) };
+            gnd_to_wikidataid_and_wikipedia_link->emplace(StringUtil::TrimWhite(line.at(0)), wiki_elements);
         }
-        file.close();
-    } else
-        LOG_ERROR("input or output files could not be opened");
+
+    } catch (const std::exception &except) {
+        LOG_ERROR("CSV to MRC has error with exception: " + std::string(except.what()));
```

Comment:
This function only covers parsing the CSV file and not writing to MRC, so the error message could instead be changed to `"Error parsing CSV file: " + ...`

Reply:
(And i am not sure whether we need to catch any exceptions anyway, since ParseCSVFileOrDie will exit the whole process on error)